Makefile.rules

```diff
@@ -0,0 +1,162 @@
+CXXFLAGS += -O2 -fPIC -Wall
+
+TEST_EXECUTABLES += $(TESTS)
+COVERAGE_EXECUTABLES += $(COVERAGE_TESTS)
+
+EXECUTABLES += $(TARGET_EXECUTABLES) $(TEST_EXECUTABLES) $(COVERAGE_EXECUTABLES)
+
+GENERATED_FILES += $(EXECUTABLES)
+
+all: $(SHARED_LIBRARIES) $(STATIC_LIBRARIES) $(TARGET_EXECUTABLES)
+
+# commands for generating various types of targets
+
+EXECUTABLE_LINK = $(CXX)
+DYNAMIC_MODULE_LINK = $(CXX) -shared
+STATIC_MODULE_LINK =$(AR) r
+
+################################################################################
+# Object build rules
+
+$(BUILD_OUT_PREFIX)%.lo: %.cpp
+	@mkdir -p $(dir $@)
+	$(CXX) -c -fPIC $(CXXFLAGS) $(CPPFLAGS) -o $@ $<
+
+$(BUILD_OUT_PREFIX)%.la: %.cpp
+	@mkdir -p $(dir $@)
+	$(CXX) -c $(CXXFLAGS) $(CPPFLAGS) -o $@ $<
+
+# static libraries
+$(BUILD_OUT_PREFIX)%.a: BASENAME=$(patsubst %.a, %, $@)
+$(BUILD_OUT_PREFIX)%.a:
+	@mkdir -p $(dir $@)
+	rm -f $@
+	ar cqv $@ $^
+	ranlib $@
+
+# dynamic libraries
+$(BUILD_OUT_PREFIX)%.so: BASENAME=$(patsubst %.so,%,$@)
+$(BUILD_OUT_PREFIX)%.so:
+	@mkdir -p $(dir $@)
+	rm -rf $@ $@.$($(BASENAME)_VERSION) $@.$($(BASENAME)_SOVERSION)
+	$(DYNAMIC_MODULE_LINK) -o $@.$($(BASENAME)_VERSION) $^ -Wl,-soname -Wl,$@.$($(BASENAME)_SOVERSION) $(LIBS)
+	ln -sf $@.$($(BASENAME)_VERSION) $@.$($(BASENAME)_SOVERSION)
+	ln -sf $@.$($(BASENAME)_VERSION) $@
+
+# binaries
+$(EXECUTABLES):
+	@mkdir -p $(dir $@)
+	$(EXECUTABLE_LINK) -o $@ $(filter %.la, $^) $(filter %.lo, $^) $(LDFLAGS) $(LIBS)
+
+################################################################################
+# Installation rules
+
+install: install-headers install-pkgconfig install-static install-shared
+
+install-headers:
+	for file in $(HEADERS) ; do \
+		$(INSTALL) -m 644 -v -D $$file $(DESTDIR)$(includedir)/$$file ; \
+	done
+
+install-pkgconfig:
+	mkdir -p $(DESTDIR)$(libdir)/pkgconfig/
+	for file in $(PKGCONFIG_FILES) ; do \
+		$(INSTALL) -m 644 -v -D $$file $(DESTDIR)$(libdir)/pkgconfig/ ; \
+	done
+
+install-static: $(STATIC_LIBRARIES)
+	mkdir -p $(DESTDIR)$(libdir)/
+	for file in $(filter %.a,$(STATIC_LIBRARIES)) ; do\
+		$(INSTALL) -m 644 -v $$file $(DESTDIR)$(libdir)/ ; \
+	done
+
+install-shared: $(patsubst %.so,install-shared-%,$(SHARED_LIBRARIES))
+install-shared-%: BASENAME=$(patsubst install-shared-%,%,$@)
+install-shared-%:
+	mkdir -p $(DESTDIR)$(libdir)/
+	$(INSTALL) -m 755 -v $(BASENAME).so.$($(BASENAME)_VERSION) $(DESTDIR)$(libdir)/
+	$(RM) -rf $(DESTDIR)$(libdir)/$(BASENAME).so $(DESTDIR)$(libdir)/$(BASENAME).so.$($(BASENAME)_SOVERSION)
+	ln -sf $(BASENAME).so.$($(BASENAME)_VERSION) $(DESTDIR)$(libdir)/$(BASENAME).so
+	ln -sf $(BASENAME).so.$($(BASENAME)_VERSION) $(DESTDIR)$(libdir)/$(BASENAME).so.$($(BASENAME)_SOVERSION)
+
+################################################################################
+# Pseudo-rules for maintaining development build tree
+
+clean:
+	find $(BUILD_OUT_PREFIX). -name "*.[ao]" -o -name "*.l[ao]" -o -name "*.so" -o -name "*.so.*" -o -name "*.gcda" -o -name "*.gcno" | xargs rm -f
```

Comment:
Why searching for all files instead of simply deleting the build directory?

Reply:
The CommandPaths specify local paths to tools that, e.g., jlc and jhls depends on, such as clang, llc, and firtool.
I was thinking that we maybe could remove it and depend on that these tools are in the user's $PATH instead.
But it's possible to have multiple versions of clang and llc installed, and I'm not sure if the version specific binary, e.g., clang-15 and llc-15, is always in the $PATH by default if only a single version of llvm is installed.

Reply:
Fixed cleaning up the generated file -- but I did not change behavior in relying on $PATH, leaving that for later if desired.